keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,569 @@
+
+### Enforcing the assignment
+
+When the workload has the PodSet assignments which are non-homogenous (all Pods
+landing in the same topology) and is about to start we modify the corresponding
+PodTemplates in the Job object to inject the `kueue.x-k8s.io/topology`
+scheduling gate. For homogenous PodSet assignments it is enough to inject the
+corresponding nodeSelector in the PodTemplates.
+
+Then, there is a new component, called `TopologyUngater`, which is a Workload
+reconciler which lists all pods for a given TAS PodSet, and ensures that the
+pods in the expected number are un-gated to a given value.
+
+Along with the scheduling gate, to each pod template the
+`kueue.x-k8s.io/workload` and `kueue.x-k8s.io/podset` labels / annotations are
+added to facilitate quick lookup (List request) for all pods corresponding to
+the workload (and more specifically PodSetAssignment) by `TopologyAssigner`.
+
+The TopologyUngater watches for pod events which trigger the reconciliation
+loop. The reconciliations are batched by 1s periods to make sure multiple
+created pods at the similar time don't trigger too many reconciliations.
+
+In order to prevent ungating more pods as expected for a PodSet we consider to
+use the expectations mechanism. The expectations are set for when we are about
+to ungate a Pod. The expectation is fulfield if the Pod is observed as ungated
+or the ungating request fails. We hold ungating if there are pending ungatings
+within the PodSet.
+
+### Support for ReplicatedJobs in JobSet
+
+Currently, in the PodSet we just keep the number of Pods created for a given
+PodTemplate. However, in case of JobSet the pods are naturally split between
+instances of replicated Jobs.
+
+Thus, without any extra work for JobSet, if JobSet has more than one
+ReplicatedJob its Pods could thus be assigned by TAS to different topologies,
+making the assignment suboptimal.
+
+In order to improve accuracy of the support for ReplicatedJobs we do the
+following API adjustments:
+
+```golang
+type PodSet struct {
+  ...
+
+  // ReplicatedJobCount indicates the number of replicated Jobs being span by
+  // the PodSet. Each replicated Job is assumed to have the same number of Pods
+  // with the same template.
+  // Default: 1
+  ReplicatedJobCount *int32
+
+  // ReplicatedJobKeyLabel specifies the name of the label which indicates
+  // the specific Job instance among ReplicatedJobs.
+  ReplicatedJobKeyLabel *string
+}
+```
+
+In case of JobSet `ReplicatedJobKeyLabel` could be either
+`jobset.sigs.k8s.io/job-index` or `jobset.sigs.k8s.io/replicatedjob-name` as
+both will uniquely identify a specific Job among the set of ReplicatedJob.
+
+```golang
+type PodSetAssignment struct {
+
+  // ReplicatedJobKey indicates the value of the Pod label which is used to
+  // specify a specific Job among a set of ReplicatedJobs.
+  ReplicatedJobKey *string
+}
+```
+Then, when the compute the assignments per pods sets corresponding to
```

Comment:
there seems to be some grammatical errors in this paragraph. Please revise.

Reply:
That's right. Rephrased.